### ISSUE TITLE
Add MultiAdam Optimizer

### ISSUE THREAD

**Issue by DanieleCucurachi** [pasqal-io/perceptrain#11]
### Describe the feature

Integration of the MultiAdam optimizer into the `perceptrain` library. MultiAdam is a scale-invariant optimizer designed for Physics-Informed Neural Networks (PINNs) that mitigates the imbalance between PDE (residual) loss term and boundary loss terms. It leverages gradient momentum to parameter-wisely balance the loss terms. The reference paper is available [here](https://arxiv.org/abs/2306.02816), and the [code is public](https://github.com/i207M/PINNacle/blob/main/src/optimizer/multiadam.py).

### It should be implemented because

Integrating MultiAdam into `perceptrain` will provide a robust optimizer tailored for PINNs, improving loss balancing in PDE-constrained optimization problems. Since the optimizer in the original code is built on top of `torch.optim.Optimizer`, it can be easily integrated into `perceptrain`.

### Additional context

A starting point for the implementation is already available in the [public repository of the paper's authors](https://github.com/i207M/PINNacle/blob/main/src/optimizer/multiadam.py). However, integrating MultiAdam into `perceptrain` comes with the following challenges:

1. Redefining the `optimize_step` function

If the core structure of the original code is maintained, which is of course convenient, the `closure` function in `perceptrain` [`optimize_step`](https://github.com/pasqal-io/perceptrain/blob/main/perceptrain/optimize_step.py) needs modification to support MultiAdam. In the original MultiAdam implementation, the optimizer calls `closure` as follows:

```python
from torch.optim import Optimizer

...

class MultiAdam(Optimizer):

...

  def step(self, closure):
          """Performs a single optimization step."""
          with torch.enable_grad():
              _ = closure(skip_backward=True)
          ...
```

A possible solution could involve modifying the `closure` function as follows (while not the most elegant solution, it effectively illustrates the required changes):
```
def closure(skip_backward=False, return_metrics=False) -> Any:
    nonlocal metrics, loss
    if not skip_backward:
        optimizer.zero_grad()
    loss, metrics = loss_fn(model, xs_to_device)
    if not skip_backward:
        loss.backward(retain_graph=True)
    if return_metrics:
        return loss.item(), metrics
    else:
        return loss.item()
```

This modification maintains backward compatibility in `perceptrain` while ensuring support for MultiAdam.

2. Handling the supervised learning data term

The original paper does not consider a data term in the PINN loss. A decision is needed on whether to group a possible data term with boundary losses, residual losses, or treat it as an independent term in the gradient-based balancing process.

3. Automating `loss_group_idx` definition

The only additional parameter that MultiAdam requires compared to standard PyTorch optimizers is `loss_group_idx`, which specifies how loss terms should be grouped for gradient balancing.

Implementing an automatic method to determine this parameter based on the specific problem being solved would significantly enhance usability and reduce manual tuning efforts.

### Would you like to work on this issue?

Yes

**Comment by RolandMacDoland** [comment]
@mlahariya Do you think this is best placed here ? That was my initial thought but might be wrong there.

**Comment by mlahariya** [comment]
> [@mlahariya](https://github.com/mlahariya) Do you think this is best placed here ? That was my initial thought but might be wrong there.

Yes, this is indeed the right place to open this issue.
